File: desktop/src-tauri/src/commands/agents.rs
//! AgentFlow commands for Phase 2E.
//!
//! Provides IPC commands for:
//! - Memory buffer management
//! - Hypothesis generation and investigation
//! - Correlation verification
//! - AgentFlow status monitoring

use serde::{Deserialize, Serialize};
use tauri::command;

// ============================================================================
// Memory Buffer Types
// ============================================================================

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemoryStatsResponse {
    pub success: bool,
    #[serde(default)]
    pub total_entries: i32,
    #[serde(default)]
    pub max_entries: i32,
    #[serde(default)]
    pub utilization: f64,
    #[serde(default)]
    pub by_type: serde_json::Value,
    #[serde(default)]
    pub by_priority: serde_json::Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemoryEntry {
    pub entry_id: String,
    pub entry_type: String,
    pub content: String,
    pub priority: String,
    pub timestamp: String,
    #[serde(default)]
    pub related_entries: Vec<String>,
    #[serde(default)]
    pub metadata: serde_json::Value,
    #[serde(default)]
    pub access_count: i32,
    pub last_accessed: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemoryEntriesResponse {
    pub success: bool,
    #[serde(default)]
    pub entries: Vec<MemoryEntry>,
    #[serde(default)]
    pub count: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemorySearchResponse {
    pub success: bool,
    pub query: Option<String>,
    #[serde(default)]
    pub entries: Vec<MemoryEntry>,
    #[serde(default)]
    pub count: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemoryClearResponse {
    pub success: bool,
    #[serde(default)]
    pub cleared_count: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

// ============================================================================
// Hypothesis Types
// ============================================================================

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Hypothesis {
    pub hypothesis_id: String,
    pub hypothesis_type: String,
    pub description: String,
    pub event_type_a: String,
    pub event_type_b: String,
    pub confidence: f64,
    #[serde(default)]
    pub evidence_for: Vec<String>,
    #[serde(default)]
    pub evidence_against: Vec<String>,
    pub status: String,
    pub created_at: String,
    pub last_updated: String,
    #[serde(default)]
    pub metadata: serde_json::Value,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HypothesesResponse {
    pub success: bool,
    #[serde(default)]
    pub hypotheses: Vec<Hypothesis>,
    #[serde(default)]
    pub count: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryPlan {
    pub hypothesis_id: String,
    #[serde(default)]
    pub queries: Vec<serde_json::Value>,
    #[serde(default)]
    pub expected_results: Vec<String>,
    #[serde(default)]
    pub completion_criteria: String,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InvestigateResponse {
    pub success: bool,
    pub hypothesis_id: Option<String>,
    pub query_plan: Option<QueryPlan>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

// ============================================================================
// Verification Types
// ============================================================================

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VerificationReport {
    pub hypothesis_id: String,
    pub result: String,
    pub confidence: f64,
    #[serde(default)]
    pub evidence_summary: String,
    #[serde(default)]
    pub criteria_met: Vec<String>,
    #[serde(default)]
    pub criteria_violated: Vec<String>,
    #[serde(default)]
    pub recommendation: String,
    pub created_at: String,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VerifyResponse {
    pub success: bool,
    pub hypothesis_id: Option<String>,
    pub report: Option<VerificationReport>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VerificationHistoryResponse {
    pub success: bool,
    pub hypothesis_id: Option<String>,
    #[serde(default)]
    pub history: Vec<VerificationReport>,
    #[serde(default)]
    pub count: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

// ============================================================================
// Status Types
// ============================================================================

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentFlowStatusResponse {
    pub success: bool,
    pub memory_buffer: Option<serde_json::Value>,
    pub correlation_planner: Option<serde_json::Value>,
    pub correlation_verifier: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportPriorsResponse {
    pub success: bool,
    #[serde(default)]
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

// ============================================================================
// Memory Buffer Commands
// ============================================================================

/// Get memory buffer statistics.
///
/// Calls: `futurnal agents memory-stats --json`
#[command]
pub async fn get_memory_stats() -> Result<MemoryStatsResponse, String> {
    let args = vec!["agents", "memory-stats", "--json"];

    let response: MemoryStatsResponse = match crate::python::execute_cli(&args).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Get memory stats CLI failed: {}", e);
            return Ok(MemoryStatsResponse {
                success: false,
                total_entries: 0,
                max_entries: 50,
                utilization: 0.0,
                by_type: serde_json::json!({}),
                by_priority: serde_json::json!({}),
                error: Some(format!("Failed to get memory stats: {}", e)),
            });
        }
    };

    log::info!("Retrieved memory stats (entries={})", response.total_entries);
    Ok(response)
}

/// Get recent memory entries.
///
/// Calls: `futurnal agents memory-recent --json [--limit N]`
#[command]
pub async fn get_memory_recent(limit: Option<i32>) -> Result<MemoryEntriesResponse, String> {
    let limit_str;
    let mut args = vec!["agents", "memory-recent", "--json"];

    if let Some(l) = limit {
        limit_str = l.to_string();
        args.push("--limit");
        args.push(&limit_str);
    }

    let response: MemoryEntriesResponse = match crate::python::execute_cli(&args).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Get memory recent CLI failed: {}", e);
            return Ok(MemoryEntriesResponse {
                success: false,
                entries: vec![],
                count: 0,
                error: Some(format!("Failed to get recent entries: {}", e)),
            });
        }
    };

    log::info!("Retrieved {} recent memory entries", response.count);
    Ok(response)
}

/// Search memory buffer for relevant entries.
///
/// Calls: `futurnal agents memory-search <query> --json [--limit N]`
#[command]
pub async fn search_memory(query: String, limit: Option<i32>) -> Result<MemorySearchResponse, String> {
    let limit_str;
    let mut args = vec!["agents", "memory-search", &query, "--json"];

    if let Some(l) = limit {
        limit_str = l.to_string();
        args.push("--limit");
        args.push(&limit_str);
    }

    let response: MemorySearchResponse = match crate::python::execute_cli(&args).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Search memory CLI failed: {}", e);
            return Ok(MemorySearchResponse {
                success: false,
                query: Some(query),
                entries: vec![],
                count: 0,
                error: Some(format!("Failed to search memory: {}", e)),
            });
        }
    };

    log::info!("Memory search found {} entries", response.count);
    Ok(response)
}

/// Clear all memory entries.
///
/// Calls: `futurnal agents memory-clear --json`
#[command]
pub async fn clear_memory() -> Result<MemoryClearResponse, String> {
    let args = vec!["agents", "memory-clear", "--json"];

    let response: MemoryClearResponse = match crate::python::execute_cli(&args).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Clear memory CLI failed: {}", e);
            return Ok(MemoryClearResponse {
                success: false,
                cleared_count: 0,
                error: Some(format!("Failed to clear memory: {}", e)),
            });
        }
    };

    log::info!("Cleared {} memory entries", response.cleared_count);
    Ok(response)
}

// ============================================================================
// Hypothesis Commands
// ============================================================================

/// List active correlation hypotheses.
///
/// Calls: `futurnal agents hypotheses --json [--status STATUS]`
#[command]
pub async fn get_hypotheses(status: Option<String>) -> Result<HypothesesResponse, String> {
    let status_owned;
    let mut args = vec!["agents", "hypotheses", "--json"];

    if let Some(s) = status {
        status_owned = s;
        args.push("--status");
        args.push(&status_owned);
    }

    let response: HypothesesResponse = match crate::python::execute_cli(&args).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Get hypotheses CLI failed: {}", e);
            return Ok(HypothesesResponse {
                success: false,
                hypotheses: vec![],
                count: 0,
                error: Some(format!("Failed to get hypotheses: {}", e)),
            });
        }
    };

    log::info!("Retrieved {} hypotheses", response.count);
    Ok(response)
}

/// Generate correlation hypotheses from event types.
///
/// Calls: `futurnal agents generate-hypotheses <event_types> --json`
#[command]
pub async fn generate_hypotheses(event_types: String) -> Result<HypothesesResponse, String> {
    let args = vec!["agents", "generate-hypotheses", &event_types, "--json"];

    let response: HypothesesResponse = match crate::python::execute_cli(&args).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Generate hypotheses CLI failed: {}", e);
            return Ok(HypothesesResponse {
                success: false,
                hypotheses: vec![],
                count: 0,
                error: Some(format!("Failed to generate hypotheses: {}", e)),
            });
        }
    };

    log::info!("Generated {} hypotheses", response.count);
    Ok(response)
}

/// Design a query plan to investigate a hypothesis.
///
/// Calls: `futurnal agents investigate <hypothesis_id> --json`
#[command]
pub async fn investigate_hypothesis(hypothesis_id: String) -> Result<InvestigateResponse, String> {
    let args = vec!["agents", "investigate", &hypothesis_id, "--json"];

    let response: InvestigateResponse = match crate::python::execute_cli(&args).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Investigate hypothesis CLI failed: {}", e);
            return Ok(InvestigateResponse {
                success: false,
                hypothesis_id: Some(hypothesis_id),
                query_plan: None,
                error: Some(format!("Failed to investigate hypothesis: {}", e)),
            });
        }
    };

    log::info!("Created investigation plan for hypothesis {}", response.hypothesis_id.as_deref().unwrap_or("unknown"));
    Ok(response)
}

// ============================================================================
// Verification Commands
// ============================================================================

/// Verify a hypothesis with current evidence.
///
/// Calls: `futurnal agents verify <hypothesis_id> --json`
#[command]
pub async fn verify_hypothesis(hypothesis_id: String) -> Result<VerifyResponse, String> {
    let args = vec!["agents", "verify", &hypothesis_id, "--json"];

    let response: VerifyResponse = match crate::python::execute_cli(&args).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Verify hypothesis CLI failed: {}", e);
            return Ok(VerifyResponse {
                success: false,
                hypothesis_id: Some(hypothesis_id),
                report: None,
                error: Some(format!("Failed to verify hypothesis: {}", e)),
            });
        }
    };

    log::info!("Verified hypothesis {}", response.hypothesis_id.as_deref().unwrap_or("unknown"));
    Ok(response)
}

/// Get verification history for a hypothesis.
///
/// Calls: `futurnal agents verification-history <hypothesis_id> --json`
#[command]
pub async fn get_verification_history(hypothesis_id: String) -> Result<VerificationHistoryResponse, String> {
    let args = vec!["agents", "verification-history", &hypothesis_id, "--json"];

    let response: VerificationHistoryResponse = match crate::python::execute_cli(&args).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Get verification history CLI failed: {}", e);
            return Ok(VerificationHistoryResponse {
                success: false,
                hypothesis_id: Some(hypothesis_id),
                history: vec![],
                count: 0,
                error: Some(format!("Failed to get verification history: {}", e)),
            });
        }
    };

    log::info!("Retrieved {} verification records", response.count);
    Ok(response)
}

// ============================================================================
// Status Commands
// ============================================================================

/// Get AgentFlow system status.
///
/// Calls: `futurnal agents status --json`
#[command]
pub async fn get_agentflow_status() -> Result<AgentFlowStatusResponse, String> {
    let args = vec!["agents", "status", "--json"];

    let response: AgentFlowStatusResponse = match crate::python::execute_cli(&args).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Get AgentFlow status CLI failed: {}", e);
            return Ok(AgentFlowStatusResponse {
                success: false,
                memory_buffer: None,
                correlation_planner: None,
                correlation_verifier: None,
                error: Some(format!("Failed to get AgentFlow status: {}", e)),
            });
        }
    };

    log::info!("Retrieved AgentFlow status");
    Ok(response)
}

/// Export AgentFlow state as natural language for token priors.
///
/// Calls: `futurnal agents export-priors --json`
#[command]
pub async fn export_token_priors() -> Result<ExportPriorsResponse, String> {
    let args = vec!["agents", "export-priors", "--json"];

    let response: ExportPriorsResponse = match crate::python::execute_cli(&args).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Export token priors CLI failed: {}", e);
            return Ok(ExportPriorsResponse {
                success: false,
                content: String::new(),
                error: Some(format!("Failed to export token priors: {}", e)),
            });
        }
    };

    log::info!("Exported token priors ({} chars)", response.content.len());
    Ok(response)
}

File: src/futurnal/cli/agents.py
"""AgentFlow CLI commands for Phase 2E.

Provides commands for:
- Memory buffer management
- Hypothesis generation and investigation
- Correlation verification
- AgentFlow status monitoring
"""

import json
import logging
from typing import Optional

from typer import Typer, Option, Argument

logger = logging.getLogger(__name__)

agents_app = Typer(help="AgentFlow analysis and correlation commands")


def _get_memory_buffer():
    """Get EvolvingMemoryBuffer singleton."""
    try:
        from futurnal.agents.memory_buffer import get_memory_buffer
        return get_memory_buffer()
    except ImportError as e:
        logger.warning(f"Memory buffer not available: {e}")
        return None


def _get_correlation_planner():
    """Get CorrelationPlanner singleton."""
    try:
        from futurnal.agents.correlation_planner import get_correlation_planner
        return get_correlation_planner()
    except ImportError as e:
        logger.warning(f"Correlation planner not available: {e}")
        return None


def _get_correlation_verifier():
    """Get CorrelationVerifier singleton."""
    try:
        from futurnal.agents.correlation_verifier import get_correlation_verifier
        return get_correlation_verifier()
    except ImportError as e:
        logger.warning(f"Correlation verifier not available: {e}")
        return None


# ============================================================================
# Memory Buffer Commands
# ============================================================================

@agents_app.command("memory-stats")
def memory_stats(
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Get memory buffer statistics.

    Examples:
        futurnal agents memory-stats
        futurnal agents memory-stats --json
    """
    try:
        buffer = _get_memory_buffer()
        if buffer is None:
            raise RuntimeError("Memory buffer not available")

        stats = buffer.get_stats()
        stats["success"] = True

        if output_json:
            print(json.dumps(stats))
        else:
            print("\nMemory Buffer Statistics")
            print("-" * 40)
            print(f"Total entries: {stats['totalEntries']}/{stats['maxEntries']}")
            print(f"Utilization: {stats['utilization']:.0%}")
            print("\nBy Type:")
            for entry_type, count in stats.get("byType", {}).items():
                print(f"  {entry_type}: {count}")
            print("\nBy Priority:")
            for priority, count in stats.get("byPriority", {}).items():
                print(f"  {priority}: {count}")

    except Exception as e:
        logger.error(f"Memory stats failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "error": str(e)}))
        else:
            print(f"Error: {e}")


@agents_app.command("memory-recent")
def memory_recent(
    limit: int = Option(10, "--limit", "-n", help="Number of entries to show"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Get recent memory entries.

    Examples:
        futurnal agents memory-recent
        futurnal agents memory-recent --limit 5 --json
    """
    try:
        buffer = _get_memory_buffer()
        if buffer is None:
            raise RuntimeError("Memory buffer not available")

        entries = buffer.get_recent(limit)

        if output_json:
            print(json.dumps({
                "success": True,
                "entries": [e.to_dict() for e in entries],
                "count": len(entries),
            }))
        else:
            print(f"\nRecent Memory Entries ({len(entries)} shown)")
            print("-" * 60)
            for entry in entries:
                ts = entry.timestamp.strftime("%Y-%m-%d %H:%M")
                print(f"[{ts}] ({entry.entry_type.value}) {entry.content[:60]}...")

    except Exception as e:
        logger.error(f"Memory recent failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "entries": [], "error": str(e)}))
        else:
            print(f"Error: {e}")


@agents_app.command("memory-search")
def memory_search(
    query: str = Argument(..., help="Search query"),
    limit: int = Option(5, "--limit", "-n", help="Maximum results"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Search memory buffer for relevant entries.

    Examples:
        futurnal agents memory-search "productivity patterns"
        futurnal agents memory-search "monday meetings" --limit 3 --json
    """
    try:
        buffer = _get_memory_buffer()
        if buffer is None:
            raise RuntimeError("Memory buffer not available")

        entries = buffer.get_relevant_context(query, max_entries=limit)

        if output_json:
            print(json.dumps({
                "success": True,
                "query": query,
                "entries": [e.to_dict() for e in entries],
                "count": len(entries),
            }))
        else:
            print(f"\nSearch Results for '{query}' ({len(entries)} found)")
            print("-" * 60)
            for entry in entries:
                ts = entry.timestamp.strftime("%Y-%m-%d")
                print(f"[{ts}] ({entry.priority.value}) {entry.content[:80]}...")

    except Exception as e:
        logger.error(f"Memory search failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "entries": [], "error": str(e)}))
        else:
            print(f"Error: {e}")


@agents_app.command("memory-clear")
def memory_clear(
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Clear all memory entries.

    Examples:
        futurnal agents memory-clear
        futurnal agents memory-clear --json
    """
    try:
        buffer = _get_memory_buffer()
        if buffer is None:
            raise RuntimeError("Memory buffer not available")

        count = buffer.clear()

        if output_json:
            print(json.dumps({"success": True, "clearedCount": count}))
        else:
            print(f"Cleared {count} memory entries")

    except Exception as e:
        logger.error(f"Memory clear failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "clearedCount": 0, "error": str(e)}))
        else:
            print(f"Error: {e}")


# ============================================================================
# Hypothesis Commands
# ============================================================================

@agents_app.command("hypotheses")
def list_hypotheses(
    status: Optional[str] = Option(None, "--status", "-s", help="Filter by status"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """List active correlation hypotheses.

    Examples:
        futurnal agents hypotheses
        futurnal agents hypotheses --status investigating --json
    """
    try:
        planner = _get_correlation_planner()
        if planner is None:
            raise RuntimeError("Correlation planner not available")

        hypotheses = planner.get_active_hypotheses()

        # Filter by status if specified
        if status:
            from futurnal.agents.correlation_planner import HypothesisStatus
            try:
                status_enum = HypothesisStatus(status)
                hypotheses = [h for h in hypotheses if h.status == status_enum]
            except ValueError:
                valid = ", ".join(s.value for s in HypothesisStatus)
                raise ValueError(f"Invalid status '{status}'. Valid: {valid}")

        if output_json:
            print(json.dumps({
                "success": True,
                "hypotheses": [h.to_dict() for h in hypotheses],
                "count": len(hypotheses),
            }))
        else:
            print(f"\nActive Hypotheses ({len(hypotheses)} total)")
            print("-" * 60)
            for h in hypotheses:
                conf = f"{h.confidence:.0%}"
                print(f"[{h.status.value}] ({conf}) {h.description[:50]}...")

    except Exception as e:
        logger.error(f"List hypotheses failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "hypotheses": [], "error": str(e)}))
        else:
            print(f"Error: {e}")


@agents_app.command("generate-hypotheses")
def generate_hypotheses(
    event_types: str = Argument(..., help="Comma-separated event types"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Generate correlation hypotheses from event types.

    Examples:
        futurnal agents generate-hypotheses "note_created,meeting_noted,decision_made"
        futurnal agents generate-hypotheses "commit_pushed,task_completed" --json
    """
    try:
        planner = _get_correlation_planner()
        if planner is None:
            raise RuntimeError("Correlation planner not available")

        types = [t.strip() for t in event_types.split(",") if t.strip()]
        if len(types) < 2:
            raise ValueError("Need at least 2 event types to generate hypotheses")

        hypotheses = planner.generate_hypotheses(types)

        if output_json:
            print(json.dumps({
                "success": True,
                "hypotheses": [h.to_dict() for h in hypotheses],
                "count": len(hypotheses),
            }))
        else:
            print(f"\nGenerated {len(hypotheses)} hypotheses")
            print("-" * 60)
            for h in hypotheses:
                print(f"- [{h.hypothesis_type.value}] {h.description}")

    except Exception as e:
        logger.error(f"Generate hypotheses failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "hypotheses": [], "error": str(e)}))
        else:
            print(f"Error: {e}")


@agents_app.command("investigate")
def investigate_hypothesis(
    hypothesis_id: str = Argument(..., help="Hypothesis ID to investigate"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Design a query plan to investigate a hypothesis.

    Examples:
        futurnal agents investigate <hypothesis_id>
        futurnal agents investigate <hypothesis_id> --json
    """
    try:
        planner = _get_correlation_planner()
        if planner is None:
            raise RuntimeError("Correlation planner not available")

        hypothesis = planner.get_hypothesis(hypothesis_id)
        if not hypothesis:
            raise ValueError(f"Hypothesis not found: {hypothesis_id}")

        plan = planner.design_query_strategy(hypothesis)

        if output_json:
            print(json.dumps({
                "success": True,
                "hypothesisId": hypothesis_id,
                "queryPlan": plan.to_dict(),
            }))
        else:
            print(f"\nQuery Plan for: {hypothesis.description[:50]}...")
            print("-" * 60)
            print(f"Queries to execute: {len(plan.queries)}")
            for i, query in enumerate(plan.queries, 1):
                print(f"  {i}. {query.get('description', query.get('type', 'unknown'))}")
            print(f"\nExpected results:")
            for result in plan.expected_results:
                print(f"  - {result}")
            print(f"\nCompletion criteria: {plan.completion_criteria}")

    except Exception as e:
        logger.error(f"Investigate failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "error": str(e)}))
        else:
            print(f"Error: {e}")


# ============================================================================
# Verification Commands
# ============================================================================

@agents_app.command("verify")
def verify_hypothesis(
    hypothesis_id: str = Argument(..., help="Hypothesis ID to verify"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Verify a hypothesis with current evidence.

    Examples:
        futurnal agents verify <hypothesis_id>
        futurnal agents verify <hypothesis_id> --json
    """
    try:
        planner = _get_correlation_planner()
        verifier = _get_correlation_verifier()

        if planner is None:
            raise RuntimeError("Correlation planner not available")
        if verifier is None:
            raise RuntimeError("Correlation verifier not available")

        hypothesis = planner.get_hypothesis(hypothesis_id)
        if not hypothesis:
            raise ValueError(f"Hypothesis not found: {hypothesis_id}")

        # Get existing evidence from hypothesis
        from futurnal.agents.correlation_verifier import EvidenceItem
        evidence = []
        for e in hypothesis.evidence_for:
            evidence.append(EvidenceItem(description=e, is_supporting=True))
        for e in hypothesis.evidence_against:
            evidence.append(EvidenceItem(description=e, is_supporting=False))

        report = verifier.verify_evidence(hypothesis, evidence)

        if output_json:
            print(json.dumps({
                "success": True,
                "hypothesisId": hypothesis_id,
                "report": report.to_dict(),
            }))
        else:
            print(f"\nVerification Report")
            print("-" * 60)
            print(report.to_natural_language())

    except Exception as e:
        logger.error(f"Verify failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "error": str(e)}))
        else:
            print(f"Error: {e}")


@agents_app.command("verification-history")
def verification_history(
    hypothesis_id: str = Argument(..., help="Hypothesis ID"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Get verification history for a hypothesis.

    Examples:
        futurnal agents verification-history <hypothesis_id>
        futurnal agents verification-history <hypothesis_id> --json
    """
    try:
        verifier = _get_correlation_verifier()
        if verifier is None:
            raise RuntimeError("Correlation verifier not available")

        history = verifier.get_verification_history(hypothesis_id)

        if output_json:
            print(json.dumps({
                "success": True,
                "hypothesisId": hypothesis_id,
                "history": [r.to_dict() for r in history],
                "count": len(history),
            }))
        else:
            print(f"\nVerification History ({len(history)} records)")
            print("-" * 60)
            for report in history:
                ts = report.created_at.strftime("%Y-%m-%d %H:%M")
                print(f"[{ts}] {report.result.value.upper()} ({report.confidence:.0%})")

    except Exception as e:
        logger.error(f"Verification history failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "history": [], "error": str(e)}))
        else:
            print(f"Error: {e}")


# ============================================================================
# Status Commands
# ============================================================================

@agents_app.command("status")
def agentflow_status(
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Get AgentFlow system status.

    Examples:
        futurnal agents status
        futurnal agents status --json
    """
    try:
        buffer = _get_memory_buffer()
        planner = _get_correlation_planner()
        verifier = _get_correlation_verifier()

        status = {
            "success": True,
            "memory_buffer": {
                "available": buffer is not None,
                "stats": buffer.get_stats() if buffer else None,
            },
            "correlation_planner": {
                "available": planner is not None,
                "active_hypotheses": len(planner.get_active_hypotheses()) if planner else 0,
            },
            "correlation_verifier": {
                "available": verifier is not None,
                "verified_count": len(verifier._verification_history) if verifier else 0,
            },
        }

        if output_json:
            print(json.dumps(status))
        else:
            print("\nAgentFlow Status")
            print("-" * 40)
            print(f"Memory Buffer: {'OK' if buffer else 'NOT AVAILABLE'}")
            if buffer:
                stats = buffer.get_stats()
                print(f"  Entries: {stats['total_entries']}/{stats['max_entries']}")
            print(f"\nCorrelation Planner: {'OK' if planner else 'NOT AVAILABLE'}")
            if planner:
                print(f"  Active hypotheses: {len(planner.get_active_hypotheses())}")
            print(f"\nCorrelation Verifier: {'OK' if verifier else 'NOT AVAILABLE'}")
            if verifier:
                print(f"  Verified hypotheses: {len(verifier._verification_history)}")

    except Exception as e:
        logger.error(f"Status check failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "error": str(e)}))
        else:
            print(f"Error: {e}")


@agents_app.command("export-priors")
def export_token_priors(
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Export AgentFlow state as natural language for token priors.

    This exports the memory buffer and correlation state as natural
    language text suitable for use as context in LLM prompts.

    Examples:
        futurnal agents export-priors
        futurnal agents export-priors --json
    """
    try:
        buffer = _get_memory_buffer()
        planner = _get_correlation_planner()
        verifier = _get_correlation_verifier()

        parts = []

        if buffer:
            parts.append(buffer.export_for_token_priors())

        if planner:
            parts.append(planner.export_for_token_priors())

        if verifier:
            parts.append(verifier.export_for_token_priors())

        content = "\n\n---\n\n".join(parts) if parts else "No AgentFlow state available."

        if output_json:
            print(json.dumps({
                "success": True,
                "content": content,
            }))
        else:
            print("\nAgentFlow Token Priors Export")
            print("=" * 60)
            print(content)

    except Exception as e:
        logger.error(f"Export priors failed: {e}")
        if output_json:
            print(json.dumps({"success": False, "content": "", "error": str(e)}))
        else:
            print(f"Error: {e}")
